addons/event_sale/models/sale_order.py

```diff
@@ -1,44 +1,32 @@
-        for order in self.browse(cr, uid, ids, context=context):
-            redirect_to_event_registration, so_id = any(line.event_id for line in order.order_line), order.id
-            order.order_line._update_registrations(confirm=True)
-        if redirect_to_event_registration:
-            event_ctx = dict(context, default_sale_order_id=so_id)
-            return self.pool.get('ir.actions.act_window').for_xml_id(cr, uid, 'event_sale', 'action_sale_order_event_registration', event_ctx)
-        else:
-            return res
+    @api.multi
+    def action_confirm(self):
+        res = super(SaleOrder, self).action_confirm()
+        if self.mapped('order_line.event_id'):
+            return self.env['ir.actions.act_window'].with_context(default_sale_order_id=self.id).for_xml_id('event_sale', 'action_sale_order_event_registration')
+        return res
 
 
-class sale_order_line(osv.osv):
+class SaleOrderLine(models.Model):
     _inherit = 'sale.order.line'
-    _columns = {
-        'event_id': fields.many2one(
-            'event.event', 'Event',
-            help="Choose an event and it will automatically create a registration for this event."),
-        'event_ticket_id': fields.many2one(
-            'event.event.ticket', 'Event Ticket',
-            help="Choose an event ticket and it will automatically create a registration for this event ticket."),
-        # those 2 fields are used for dynamic domains and filled by onchange
-        # TDE: really necessary ? ...
-        'event_type_id': fields.related('product_id', 'event_type_id', type='many2one', relation="event.type", string="Event Type"),
-        'event_ok': fields.related('product_id', 'event_ok', string='event_ok', type='boolean'),
-    }
 
-    def _prepare_order_line_invoice_line(self, cr, uid, line, account_id=False, context=None):
-        res = super(sale_order_line, self)._prepare_order_line_invoice_line(cr, uid, line, account_id=account_id, context=context)
+    event_id = fields.Many2one('event.event', 'Event', help="Choose an event and it will automatically create a registration for this event.")
+    event_ticket_id = fields.Many2one('event.event.ticket', 'Event Ticket', help="Choose an event ticket and it will automatically create a registration for this event ticket.")
+    event_type_id = fields.Many2one(related='product_id.event_type_id', string="Event Type")
+    event_ok = fields.Boolean(related='product_id.event_ok')
+
+    @api.model
+    def _prepare_order_line_invoice_line(self, line, account_id=False, context=None):
+        res = super(SaleOrderLine, self)._prepare_order_line_invoice_line(line, account_id=account_id)
         if line.event_id:
-            event = self.pool['event.event'].read(cr, uid, line.event_id.id, ['name'], context=context)
-            res['name'] = '%s: %s' % (res.get('name', ''), event['name'])
+            res['name'] = '%s: %s' % (res.get('name', ''), line.event_id.name)
         return res
 
     @api.onchange('product_id')
```

Comment:
you may change the code of onchange to something like,

--
        values = {'event_type_id': False, 'event_ok': False}
        if self.product_id.event_ok:
            values.update(event_type_id=self.product_id.event_type_id.id,
                          event_ok=self.product_id.event_ok)
        ....

--